Hook up the new frameworks with Actionable Error

gsamokovarov opened the pull request:
The idea behind actionable errors tries to help with the development experience at a start of a project or while trying to setup or test a new framework. I have introduced actionable errors if you try to use new frameworks, but forgot set them up.

<details>
  <summary>Active Storage (GIF example)</summary>

  ![rails_active_storage](https://user-images.githubusercontent.com/604618/59213990-d085fd80-8bbe-11e9-8a77-fefa3b96cd28.gif)
</details>

<details>
  <summary>Action Mailbox (GIF example)</summary>

  ![rails_action_mailbox_optimized](https://user-images.githubusercontent.com/604618/59214217-50ac6300-8bbf-11e9-9721-d16842978475.gif)
</details>

<details>
  <summary>Action Text (GIF example)</summary>

  ![rails_action_text](https://user-images.githubusercontent.com/604618/59214234-5d30bb80-8bbf-11e9-9901-c433962fc600.gif)
</details>

The current flow works, but we may see problems running possibly interactive commands like `yarn:install` from the Action Text actionable error.

Now that we have actionable errors for all internal frameworks, if we want plugin developers to use them, have to be easy to define. Setting up the internal components gave us a common use case that we try to optimize for in this PR.

You can hook up an actionable error on the exception that crashes the application when you try to use the framework, without first setting it up.

```ruby
class ActiveStorage::InstallError < Error
  include ActiveSupport::ActionableError

  def initialize(message = nil)
    super(message || <<~MESSAGE)
      Action Mailbox does not appear to be installed. Do you want to install it now?
    MESSAGE
  end

  trigger on: ActiveRecord::StatementInvalid, if: -> error do # 👈
    error.message.match?(InboundEmail.table_name)
  end

  action "Install now" do
    Rails::Command.invoke("active_storage:install")
    Rails::Command.invoke("db:migrate")
  end
end
```


The trigger says: look for unhandled `ActiveRecord::StatementInvalid` errors and reraise the current actionable error (without any arguments to the its constructor, e.g. expect a default error message). The actionable error is risen by the `ActionDispatch::ActionableExceptions` middleware, which is already in the default middleware stack.

This makes the actionable errors self-contained – the code for the action and how they are triggered is in one place!

~There are still some gotchas, though. You may want to define the error only if the `Rails::Railtie` are defined, as you may invoke commands with `Rails::Command`. In all the internal gems, we try to define the errors only if the gems are used with Rails, and not outside of it, as all of them (`activestorage`, `actionmailbox`, `actiontext`) do not explicitly depend on `railties`.~

~It may make sense to define the actionable errors in the railties with if this becomes a problem.~

<details>
  <summary>Example of Actionable Error definition in a Railtie</summary>

```ruby
# WARNING: THIS IS HYPOTHETICAL! 😅
initializer "active_storage.actionable_errors" do
  ActiveSupport::ActionableError.define :InstallError, under: ActiveStorage do
    message <<~END
      Action Mailbox does not appear to be installed. Do you want to install it now?
    END

    trigger on: ActiveRecord::StatementInvalid, if: -> error do # 👈
      error.message.match?(InboundEmail.table_name)
    end

    action "Install now" do
      Rails::Command.invoke("active_storage:install")
      Rails::Command.invoke("db:migrate")j
    end
  end
end
```

</details>

## Known Problems

- Action Text install does not find the yarn.lock file.
~- The output capture can create issues. I've seen action runs with only partial information.~

gsamokovarov committed:
Trigger Actionable Error from regular Exceptions

Define an Actionable Error trigger with:

```ruby
class ActionMailbox::InstallError
  include ActiveSupport::ActionableError

  trigger on: ActiveRecord::StatementInvalid, if: -> error do
    error.message.match?(InboundEmail.table_name)
  end
end
```

The `ActionDispatch::ActionableExceptions` middleware knows about
`ActiveSupport::ActionableErrors` triggers and can reraise an actionable
error if the current unhandled exception matches a trigger condition.

gsamokovarov committed:
Setup Active Storage with an Actionable Error

gsamokovarov committed:
Setup Action Mailbox with an Actionable Error

gsamokovarov committed:
Introduce custom template for Actionable Errors in DebugExceptions

gsamokovarov committed:
Setup Action Text with an Actionable Error

paracycle commented on activesupport/CHANGELOG.md:
```diff
@@ -1,3 +1,34 @@
+*   Trigger `ActiveSupport::ActionableError` from existing errors.
+
+    Triggers let's you re raise an actionable error from an existing one, if it
```
`re-raise` since it should be one word.

gsamokovarov replied:
Thanks! Fixed it.

paracycle reviewed (approved):
Looks great to me!

rafaelfranca commented on actionmailbox/lib/action_mailbox/errors.rb:
```diff
@@ -0,0 +1,32 @@
+  # Generic base class for all Action Mailbox exceptions.
+  class Error < StandardError; end
+
+  if defined?(Rails)
```
This check tells me that this would be better in its own file loaded by the railtie.

gsamokovarov replied:
I have gone with Kasper's API idea approach and defined the actionable errors straight in the railties. WDYT?

rafaelfranca commented on actionpack/lib/action_dispatch/middleware/exception_wrapper.rb:
```diff
@@ -55,7 +55,11 @@ def unwrapped_exception
     def rescue_template
-      @@rescue_templates[@exception.class.name]
+      @@rescue_templates.each do |error_class, template|
+        return template if error_class.safe_constantize === @exception
```
This is a breaking change, isn't? Now the first exception that get matched will take precedence over others.

gsamokovarov replied:
It is, actually. I haven't looked at the currently registered errors, but the hierarchy check may introduce changes. Will take a look.

gsamokovarov committed:
Introduce `ActiveSupport::ActionableError.define`.

This can be useful if you're code does not depend strictly on Rails. In
this case, you may not want to define an error depending on Rails
specific code next to your regular error class hierarchy.

kaspth commented on actionmailbox/lib/action_mailbox/engine.rb:
```diff
@@ -29,5 +30,22 @@ class Engine < Rails::Engine
     end
+
+    initializer "action_mailbox.actionable_errors" do
+      ActiveSupport::ActionableError.define :InstallError, under: ActionMailbox do |actionable|
```
Still think these should be `:MissingInstallError` since there's not been an error when installing it.

kaspth commented on actionmailbox/lib/action_mailbox/engine.rb:
```diff
@@ -29,5 +30,22 @@ class Engine < Rails::Engine
+          Rails::Command.invoke("db:migrate")
+        end
+      end
+    end
```
Digging `define`! Nicely solves the autoloading issues.

kaspth commented on actionpack/lib/action_dispatch/middleware/actionable_exceptions.rb:
```diff
@@ -16,24 +16,54 @@ def call(env)
+        origin_stream = stream_io.dup
+        stream_io.reopen(captured_stream)
+
+        yield
```
How are we actually capturing `$stderr` since the result from yield here isn't flowing up to `capture :stdout` far as I can see?

Don't we actually want stdout and stderr to share the same Tempfile io? Aren't the messages out of order otherwise? And does it have to be a Tempfile or is a plain `StringIO.new` good enough?

Would this work?

```ruby
def capture_ios
  ios = [ $stdout, $stderr ]
  $stdout = $stderr = shared_io = StringIO.new

  yield

  shared_io.string
ensure
  $stdout, $stderr = ios
end
```

gsamokovarov replied:
The `capture_ios` will capture output issued from the current process (say we invoke a method in the action and it `puts`), but won't caputre output from child processes like the ones created during `system` calls. We call `Rails::Command.invoke` that invokes rake tasks that usually start other processes. You have to `reopen` the streams for capturing the chile process output as the assignment does not propagate the streams to the children, but keeps the "original" ones.

I have spent more time on a capturing implementation that is good with both child capture and current process output handling.

kaspth commented on actionpack/lib/action_dispatch/middleware/templates/rescues/actionable_error.html.erb:
```diff
@@ -0,0 +1,65 @@
+  <h1>
+    <%= @exception.class.to_s %>
+    <% if @request.parameters['controller'] %>
+      in <%= @request.parameters['controller'].camelize %>Controller<% if @request.parameters['action'] %>#<%= @request.parameters['action'] %><% end %>
```
Can this be simplified to? `[ "#{@request.parameters['controller'].camelize}Controller", @request.parameters['action'] ].compact.join("#") %>`

kaspth commented on actionpack/lib/action_dispatch/middleware/templates/rescues/actionable_error.html.erb:
```diff
@@ -0,0 +1,65 @@
+
+<div id="container">
+  <h2>
+    <%= h @exception.message %>
```
Why do we need `html_escape` here? I thought Rails did it automatically?

gsamokovarov replied:
Sorry for the late chime-in. I copied the template from the other and haven't given it a thought. 😅 Could be a "historical" (not to be confused with legacy) reason. I will check!

kaspth commented on actionpack/lib/action_dispatch/middleware/templates/rescues/actionable_error.html.erb:
```diff
@@ -0,0 +1,65 @@
+</div>
+
+<script type="text/javascript">
+  document.addEventListener("submit", function(event) {
```
Why did we need to do an Ajax submit here again?

kaspth commented on actionpack/lib/action_dispatch/middleware/templates/rescues/actionable_error.html.erb:
```diff
@@ -0,0 +1,65 @@
+
+    var internalError = false;
+
+    fetch(url, {method: "POST", body: formData})
```
If we have access to `fetch`, do we have access to native ES5/ES6?

kaspth commented on activestorage/lib/active_storage/engine.rb:
```diff
@@ -143,5 +144,22 @@ class Engine < Rails::Engine # :nodoc:
+    initializer "active_storage.actionable_errors" do
+      ActiveSupport::ActionableError.define :InstallError, under: ActiveStorage do |actionable|
+        actionable.message <<~MESSAGE
+          Action Storage does not appear to be installed. Do you want to install it now?
```
`Active Storage`

kaspth commented on activesupport/CHANGELOG.md:
```diff
@@ -1,3 +1,66 @@
+*   Introduce `ActiveSupport::ActionableError.define`.
+
+    This can be useful if you're code does not depend strictly on Rails. In
```
`your`. Should this say Rails or just Active Support? It's a weird boundary that we mention Rails in the second line even though this is an Active Support feature. So maybe we should mention how to catch and triggers these in people's own apps?

kaspth replied:
Riiiight, we're trying to allude to that this depends on the Action Pack raise/rescue to get built in support.

kaspth commented on activesupport/CHANGELOG.md:
```diff
@@ -1,3 +1,66 @@
+    this case, you may not want to define an error depending on Rails specific
+    code next to your regular error class hierarchy.
+
+    You can define it the actionable error straight in a Railtie, next to you
```
```suggestion
    You can define the actionable error in a Railtie, next to your
```

kaspth commented on activesupport/CHANGELOG.md:
```diff
@@ -1,3 +1,66 @@
+
+*   Trigger `ActiveSupport::ActionableError` from existing errors.
+
+    Triggers let's you re raise an actionable error from an existing one, if it
```
```suggestion
    Triggers lets you re-raise an actionable error from an existing one, if it
```

Lets without apostrophe because let's expands to "let us".

kaspth commented on activesupport/CHANGELOG.md:
```diff
@@ -1,3 +1,66 @@
+      include ActiveSupport::ActionableError
+
+      def initialize(message = nil)
+        super(message || <<~MESSAGE)
```
I still don't understand why we do `message ||`. Why do we need the default?

kaspth commented on activesupport/lib/active_support/actionable_error.rb:
```diff
@@ -43,6 +135,27 @@ module ClassMethods
+      # Note: For an actionable error to be triggered, its constant needs to be
+      # loaded. If it isn't autoloaded, the triggers won't be registered until
+      # it is. Keep that in mind, if the actionable error isn't triggered.
+      def trigger(options)
```
```suggestion
      def trigger(on:, if:)
```

Gets the keyword errors for free.

gsamokovarov replied:
I cannot use `if` as a local variable though 🤷‍♂️

kaspth replied:
That's what `binding.local_variable_get(:if)` in my other suggestion is for.

gsamokovarov replied:
Sure, I was trying to avoid the workarounds. Here it is, though.

kaspth commented on activesupport/lib/active_support/actionable_error.rb:
```diff
@@ -43,6 +135,27 @@ module ClassMethods
+        error = options.fetch(:on) { raise ArgumentError, "missing keyword: on" }
+        condition = options.fetch(:if) { raise ArgumentError, "missing keyword: if" }
+
+        ActiveSupport::ActionableError.triggers[error] << Trigger.new(self, condition)
```
```suggestion
        ActiveSupport::ActionableError.triggers[on] << Trigger.new(self, binding.local_variable_get(:if))
```

kaspth commented on activesupport/lib/active_support/actionable_error.rb:
```diff
@@ -9,12 +9,99 @@ module ActiveSupport
+
+    Trigger = Struct.new(:actionable, :condition) do # :nodoc:
+      def act_on(error)
+        raise actionable if condition&.call(error)
```
```suggestion
        raise actionable if condition.call(error)
```

Condition is required. It'll never be nil.

kaspth commented on activesupport/lib/active_support/actionable_error.rb:
```diff
@@ -9,12 +9,99 @@ module ActiveSupport
+            action(name, &block)
+          end
+
+          trigger(actionable.trigger) unless actionable.trigger.empty?
```
Does it make sense to have one of these without a trigger?

gsamokovarov replied:
Actuall, maybe no if you're defining it with the DSL.

kaspth commented on activesupport/lib/active_support/actionable_error.rb:
```diff
@@ -9,12 +9,99 @@ module ActiveSupport
+    #    end
+    #  end
+    def self.define(name, under:, &block)
+      definition = Definition.new(&block)
```
Just inline this variable.

kaspth reviewed (commented):
Pretty close!

kaspth commented:
Looks like there's also some warnings we want to address:

```
/rails/activesupport/lib/active_support/actionable_error.rb:78: warning: instance variable @actions not initialized
/rails/activesupport/lib/active_support/actionable_error.rb:66: warning: instance variable @trigger not initialized
/rails/activesupport/lib/active_support/actionable_error.rb:56: warning: instance variable @message not initialized
```

gsamokovarov committed:
Overhaul the actionable errors output capturing

The current approach deals better both with capturing of output issued
by the current process and with output issued by child processes (system
calls and the likes).

gsamokovarov committed:
Clean up the actionable errors template

gsamokovarov commented:
@kaspth I have fixed the warning and improved the output capturing. Mind taking another look?

kaspth commented:
Great! I dig the capture output change you made. I don't think I have more feedback than what I added last time, there's some spelling things — I can go without the ES6 suggestion too and the Ajax question.

@rafaelfranca seemed to catch a breaking change, but I don't know if it actually was? https://github.com/rails/rails/pull/36455#discussion_r293089087

If so, that's the only real blocker that I have. Great job, Genadi! ❤️

tbuehlmann commented on activesupport/CHANGELOG.md:
```diff
@@ -1,3 +1,66 @@
+
+      action "Install now" do
+        Rails::Command.invoke("active_storage:install")
+        Rails::Command.invoke("db:migrate")j
```
The `j` here is probably a typo?

gsamokovarov replied:
It is, thanks for catching!

kaspth commented:
I’ll be taking a look at this 👍
